tools/break.sh

```diff
@@ -0,0 +1,41 @@
+#!/bin/sh
+# Shell script equivalent of Metamath BREAK tool
+# Mario Carneiro 7-Jan-2022
+
+usage() {
+  cat << HELP
+BREAK - Break up (parse) a file into a list of tokens (one per line)
+
+This command breaks up a file into tokens, one per line, breaking at
+whitespace and any special characters you specify as delimiters.
+Use an explicit (quoted) space as SPECCHARS to avoid the default
+special characters and break only on whitespace.
+Syntax:  break.sh SPECCHARS < INFILE > OUTFILE
+HELP
+}
+
+if [ $# -ne 1 ]; then usage; exit 1; fi
+
+if [ "$1" = "" ]; then specChars="()[],=:;{}"; else specChars=$1; fi
+
+awk -v specChars="$specChars" \
+'BEGIN {
+  len = length(specChars);
+  split(specChars, specCharsArr, "");
+} {
+  for (i = 1; i <= len; i++) {
+    if (!index($0, specCharsArr[i])) continue;
```

Comment:
I think the implementation of the loop works. But is
```
escaped = specCharsArr[i];
sub(/[\\\"]/, "\\" specCharsArr[i], escaped);
sub(escaped, "  " specCharsArr[i] " ");
```
not more effective?

Reply:
You would still have to traverse over the matched character to avoid an infinite loop. I believe that something like this can work, but it's not obvious to me that it is shorter. It would be nice if there were a way to avoid the loop too and do it all with one regex.

Reply:
the 3rd line replaces most of your while loop. I don't want to be nitpicky here. It is several years ago that I wrote my last awk script, and I forgot 80% or more about the syntax and efficient coding patterns. Eventually I noticed that I never saw the match shorthand & in replacement strings, and here is an example of how to use it.

Reply:
This eliminates the whole for loop and makes the split command in BEGIN redundant
```
    escaped = specChars;
    # metachars need to be escaped to work as a dynamic regexp
    gsub(/[\\\]\-\^]/, "\\\\&", escaped);
    escaped = "[" escaped "]";
    print escaped;
    gsub(escaped, " & ");
```

Reply:
updated

Reply:
ack